tianshou/policy/base.py

```diff
@@ -201,51 +226,72 @@ def compute_nstep_return(
-        gammas = np.zeros_like(indice) + n_step
-        done, buf_len = buffer.done, len(buffer)
-        for n in range(n_step - 1, -1, -1):
-            now = (indice + n) % buf_len
-            gammas[done[now] > 0] = n
-            returns[done[now] > 0] = 0
-            returns = (rew[now] - mean) / std + gamma * returns
+            mean, std = 0., 1.
+        buf_len = len(buffer)
         terminal = (indice + n_step - 1) % buf_len
         target_q_torch = target_q_fn(buffer, terminal).flatten()  # (bsz, )
         target_q = to_numpy(target_q_torch)
-        target_q[gammas != n_step] = 0
-        target_q = target_q * (gamma ** gammas) + returns
+
+        target_q = _nstep_return(rew, buffer.done, target_q, indice,
+                                 gamma, n_step, len(buffer), mean, std)
+
         batch.returns = to_torch_as(target_q, target_q_torch)
         # prio buffer update
         if isinstance(buffer, PrioritizedReplayBuffer):
             batch.weight = to_torch_as(batch.weight, target_q_torch)
         return batch
 
-    def post_process_fn(self, batch: Batch,
-                        buffer: ReplayBuffer, indice: np.ndarray) -> None:
-        """Post-process the data from the provided replay buffer. Typical
-        usage is to update the sampling weight in prioritized experience
-        replay. Check out :ref:`policy_concept` for more information.
-        """
-        if isinstance(buffer, PrioritizedReplayBuffer) \
-                and hasattr(batch, 'weight'):
-            buffer.update_weight(indice, batch.weight)
 
-    def update(self, batch_size: int, buffer: Optional[ReplayBuffer],
-               *args, **kwargs) -> Dict[str, Union[float, List[float]]]:
-        """Update the policy network and replay buffer (if needed). It includes
-        three function steps: process_fn, learn, and post_process_fn.
+@njit
+def _episodic_return(
+    v_s_: np.ndarray, rew: np.ndarray, done: np.ndarray,
+    gamma: float, gae_lambda: float,
+) -> np.ndarray:
+    """Numba speedup: 4.1s -> 0.057s"""
+    returns = np.roll(v_s_, 1)
+    m = (1. - done) * gamma
+    delta = rew + v_s_ * m - returns
+    m *= gae_lambda
+    gae = 0.
+    for i in range(len(rew) - 1, -1, -1):
+        gae = delta[i] + m[i] * gae
+        returns[i] += gae
+    return returns
+
+
+@njit
+def _nstep_return(
+    rew: np.ndarray, done: np.ndarray, target_q: np.ndarray,
+    indice: np.ndarray, gamma: float, n_step: int, buf_len: int,
+    mean: float, std: float
+) -> np.ndarray:
+    """Numba speedup: 0.3s -> 0.15s"""
+    returns = np.zeros(indice.shape)
+    gammas = np.full(indice.shape, n_step)
+    for n in range(n_step - 1, -1, -1):
+        now = (indice + n) % buf_len
+        gammas[done[now] > 0] = n
+        returns[done[now] > 0] = 0.
+        returns = (rew[now] - mean) / std + gamma * returns
+    target_q[gammas != n_step] = 0
+    target_q = target_q * (gamma ** gammas) + returns
+    return target_q
+
+
+def _compile():
```

Comment:
Why do you want to precompile the function ? I don't see the point. It goes against the JIT (Just-in-Time compilation) principle. I would use instead AOT (Ahead-of-Time compilation) if you really think it is better to pre-compile whose methods.

Here is the [official documentation](https://numba.pydata.org/numba-doc/dev/user/pycc.html), including examples.

Reply:
I post the reason in the docstring. Since I want to compare the training time of several versions, if we do not compile the main function pattern ahead of time, the first-time compile-time will add onto the training time in the collector, and thus it will show that Numba slows down the whole process (from 2050 fps to 1800 fps).

Reply:
I don't think AOT is a good choice since we cannot determine some arguments' type ahead of time, for example, `v_s_` in GAE.

Reply:
Then I would rather move _compile at the beginning of the brenchmark, rather than here. I thing it makes more sense.

Reply:
> I don't think AOT is a good choice since we cannot determine some arguments' type ahead of time, for example, v_s_ in GAE.

I don't now, it is strange. Compilation support variable size numpy array, just the number of dimensions must be known in advance. Knowing that you can also precompile several signatures if necessary.